docs/root/configuration/http_filters/wasm_filter.rst

```diff
@@ -0,0 +1,1184 @@
+                        StringView service_name,
+                        StringView method_name, 
+                        const google::protobuf::MessageLite &request, 
+                        uint32_t timeout_milliseconds,
+                        std::function<void(Response&& response)> success_callback,
+                        std::function<void(GrpcStatus status, StringView error_message)> failure_callback)
+
+Makes a unary gRPC call to an upstream host.
+
+*service* is a serialized proto string of :ref:`gRPC service <envoy_api_msg_core.GrpcService>` for gRPC client initialization.
+*service_name* and *method_name* indicates the target gRPC service and method name.
+*request* is a `lite proto message <https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.message_lite>`_ that gRPC service accepts as request.
+*timeout_milliseconds* is an unsigned integer as timeout period for the gRPC call in milliseconds.
+*success_callback* is the callback function that will be called when gRPC call succeeds. *response* is the returned message from gRPC service.
+*failure_callback* is the callback function that will be invoked when gRPC call fails. *status* is the returned gRPC status code. *error_message* is detailed error message extracted from gRPC response.
+
+.. note::
+    if the call outlives the stream context, *grpcSimpleCall* should be called within :ref:`root context <config_http_filters_wasm_context_object>`.
+
+grpcCallHandler
+^^^^^^^^^^^^^^^
+
+.. code-block:: cpp
+
+    void grpcCallHandler(
+        StringView service,
+        StringView service_name,
+        StringView method_name,
+        const google::protobuf::MessageLite &request,
+        uint32_t timeout_milliseconds,
+        std::unique_ptr<GrpcCallHandlerBase> handler)
+
+Makes a unary gRPC call to an upstream host.
+
+Similar to :ref:`grpcSimpleCall <config_http_filters_wasm_context_object_api_grpcSimpleCall>` for gRPC client initialization,
+but uses :ref:`GrpcCallHandler <config_http_filters_wasm_GrpcCallHandler>` as target for callback and fine grained control on the call.
+
+grpcStreamHandler
+^^^^^^^^^^^^^^^^^
+
+.. code-block:: cpp
+
+    void grpcStreamHandler(StringView service,
+                           StringView service_name,
+                           StringView method_name,
+                           std::unique_ptr<GrpcStreamHandlerBase> handler)
+
+Makes an gRPC stream to an upstream host.
+
+*service* is a serialized proto string of :ref:`gRPC service <envoy_api_msg_core.GrpcService>` for gRPC client initialization.
+*service_name* and *method_name* indicates the target gRPC service and method name.
+*handler* (:ref:`GrpcStreamHandler <config_http_filters_wasm_GrpcStreamHandler>`) is used to control the stream and
+as target for gRPC stream callbacks.
+
+.. note::
+    if the stream call outlives the per request context, *grpcStreamHandler* should be called within :ref:`root context <config_http_filters_wasm_context_object>`.
```

Comment:
This is an issue.  There is no easy way to create a callback on another context_id.  Moreover onDelete will delete the Context making it impossible to get the gRPC result.   We could decouple the gRPC upcalls from the stream context and always invoke them with the root context id.  The high level code would then invoke a std::function which could do whatever, including attempt to find a context with (a new) getContext(stored_context_id) call.   Alternatively I can block the doDelete call until the last http/gRPC callback has arrived... thoughts?

Reply:
I think it makes sense to hide this in API and make calls from stream context always attach to root context.

Reply:
For this PR though I think we could leave it as what it is and remove the notes when the improvement is made? I can create a issue for this.

Reply:
https://github.com/envoyproxy/envoy-wasm/issues/85 Issue created.

Reply:
LGTM